### DISCUSSION TITLE
What is the sector ordering / skew for Apple ][ Pascal disks and CP/M disks?

### DISCUSSION THREAD

**Discussion by hippietrail** [fadden/CiderPress2#37] [Q&A] [answered]
Writing my own code for fun I've got DOS 3 and ProDOS sector ordering working.

But I started looking at .dsk image files of Pascal and CP/M disks, and Googling about their sector ordering and/or skew and I'm not finding much.

I downloaded a random set of each and I've played with CP/M images for other systems a little bit but not never looked at Apple II Pascal disk images before. It seems like all the CP/M ones have their sectors in the same order and all the Pascal ones have theirs in the same order. All the file extensions are `.dsk` or `.do`.

The Pascal images all have the directory/catalog starting at offset `$0b00` so 'image sector' 11 of track 0. I was thinking of this as 'physical sector 11' but that doesn't seem to be accurate. Because when I open the image in AppleSauce and order by 'physical sectors' it's at sector 8 or track 0.

The CP/M images all have the directory/catalog starting at offset `$03000` so 'image sector' 0 of track 3.

But the second sector of the directory is not the next sector of the disk image in either case.

The second sector of the dir in the Pascal images is at 'physical sector' 10 of track 0. The second sector of the dir in the CP/M images is at 'physical sector 3 of track 3'.

In 'disk image order' the second sector of the Pascal dir is at `$0a00` or sector 10 of track 0 - same as physical order, but the second sector of the CP/M dir is at `$3600` so sector 6 of track 3 - different to physical order.

So I feel I must be missing something that's difficult to Google that's different than DOS 3 and ProDOS disk images. Either instead of or in addition to the old do/po sector ordering. Is this sector skew? How do I get my head around it? And do these also come in .po/ProDOS ordering too?

**Comment by rb6502** [comment]
.DSK images don't have a defined sector ordering.  Some are DOS 3.3 order, some are ProDOS order.  The actual OS format doesn't matter and I've seen Pascal and CP/M .DSKs both ways.  Yes, that's terrible, but it's an almost 30 year precedent now.

**Comment by fadden** [comment, marked as answer]
The ".do" or ".po" is a reflection of the operating system in use by the tool that created the disk image.  DOS tools like DDD create DOS-ordered images, while ProDOS tools like ShrinkIt create ProDOS-ordered images.  Emulator tools do one or the other.  This is independent of the sector skew.

UCSD Pascal disks - which can be found for other systems (e.g. https://retrocomputing.stackexchange.com/a/28487/56) - use ProDOS sector skewing.  Technically, Pascal came first, so it might be more accurate to say that ProDOS disks use Pascal skewing.

Apple II CP/M disks have their own sector skewing... but the first few tracks are usually in ProDOS order.

Sector skewing for DOS, Pascal, and CP/M was first(?) documented in Beneath Apple DOS: https://archive.org/details/Beneath_Apple_DOS_alt/page/n17/mode/1up .  The skew translation tables are in code form [here](https://github.com/fadden/CiderPress2/blob/232324a3f15f9d32c230558d2d4fe8f65787e92a/DiskArc/GeneralChunkAccess.cs#L50).

https://ciderpress2.com/doc-index.html has pages for Pascal and CP/M filesystems.  Pascal uses 512-byte blocks, same as ProDOS, while CP/M on 5.25" disks uses 1024-byte blocks.  CP/M data disks can actually wrap around at the end.

https://github.com/fadden/CiderPress2/tree/main/DiskArc/FS has the CP2 implementations.  There's a function called `TestImage` in the main class for each filesystem that does the filesystem recognition, e.g. a pretty simple one for [Pascal](https://github.com/fadden/CiderPress2/blob/232324a3f15f9d32c230558d2d4fe8f65787e92a/DiskArc/FS/Pascal.cs#L176) and this rather lengthy function for [CP/M](https://github.com/fadden/CiderPress2/blob/232324a3f15f9d32c230558d2d4fe8f65787e92a/DiskArc/FS/CPM.cs#L237).

FWIW, the sector editor in CiderPress II allows you to change the sector skew currently in use:
![image](https://github.com/user-attachments/assets/1229905b-899f-42da-9133-2a1744c25fe3)
(The pop-up should actually be labeled "sector skew" to avoid confusion with the disk image ordering.)

**Reply by hippietrail** [reply]
I was actually getting the feeling Pascal used ProDOS skew but was thrown off by a) Just doing DOS 3.3 <-> ProDOS reordering, I don't actually see ProDOS skew and b) the anachronism made it feel wrong - so thanks for confirming that Pascal did use it first!

I have used your docs a lot to get this far and still have those three CiderPress source files open in other tabs and though they helped with the filesystem formats, I didn't manage to glean the sector ordering/skew.

I don't have access to Windows these days but have been carrying around my old Intel MacBook for years with the intention of installing Windows on it. I just foresee too much frustration to have tried it yet (-:
